include/ginkgo/core/base/mpi.hpp

```diff
@@ -529,11 +529,31 @@ class communicator {
     template <typename SendType>
     request i_send(const SendType* send_buffer, const int send_count,
                    const int destination_rank, const int send_tag) const
+    {
+        return i_send(send_buffer, send_count, type_impl<SendType>::get_type(),
+                      destination_rank, send_tag);
+    }
+
+    /**
+     * Send (Non-blocking, Immediate return) data from calling process to
+     * destination rank.
+     *
+     * @param send_buffer  the buffer to send
+     * @param send_count  the number of elements to send
+     * @param send_type  the MPI_Datatype of the elements to send
+     * @param destination_rank  the rank to send the data to
+     * @param send_tag  the tag for the send call
+     *
+     * @return  the request handle for the send call
+     */
+    request i_send(const void* send_buffer, const int send_count,
+                   MPI_Datatype send_type, const int destination_rank,
```

Comment:
I guess you want to mainly use this is for sending custom datatypes ? Like packed data ?

Reply:
I don't think I actually use the `i_send` version. For `i_all_to_all_v` it's necessary for multiple rhs in our SpMV. But I think I can remove the `i_send` (and `i_recv`) overloads. But I will do that in the #1054 because rebasing would be annoying.